### PR TITLE
Display DatasetSelector warning when file(s) don't exist.

### PR DESCRIPTION
**Pull request by dustins**
*Issue #, if available:*

*Description of changes:*

**Changes**
- (frontend) Selectable types for the DatasetSelector is now tied to the S3 data type. Selecting `S3Prefix` will constraint the DatasetSelector to prefixes whereas `ManifestFile` or `AugmentedManifestFile` will force object selection.

- (frontend) When the `DatasetSelector` is used to select a file or prefix from a Dataset we know that a user wants to select a file or files respectively. This adds a warning if you're selecting a file and it does not exist or a prefix and no files exist under that prefix. <br /> <br /> Screenshot 2024-04-24 at 14 09 32

- (backend) Updated Dataset list_files lambda to allow defining the s3 delimiter via query string parameter. By default `/` is used so only files/prefixes that are direct descendants of the prefix are returned. The change allows querying for any file starting with a prefix no matter how deeply it is nested.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: frontend/src/modules/dataset/dataset-selector.tsx [hunk 1]
```diff
@@ -65,7 +121,22 @@ export function DatasetResourceSelector (props: DatasetResourceSelectorProps) {
         <>
             <FormField {...fieldProps}>
                 <Grid gridDefinition={[{colspan: 6}, {colspan: 6}]}>
-                    <Input placeholder={s3UriPlaceholder} {...inputProps} type='search'></Input>
+                    <SpaceBetween direction='vertical' size='xs'>
+                        <Input placeholder={s3UriPlaceholder} {...inputProps} type='search'></Input>
+                        <Condition condition={!!props.alertOnEmpty && state.isEmpty}>
+                            <Alert statusIconAriaLabel='Warning' type='warning'>
+                                <Condition condition={isSelectingObject && isSelectingPrefixes}>
+                                        No file found with this name and no files found with this prefix.
```
Comment (KyleLilly) [inline frontend/src/modules/dataset/dataset-selector.tsx:129]: Seems a little wordy. Can we just make it something like this or do you think that's missing context?
```suggestion
                                        No file found with this name or prefix.
```

**Review by KyleLilly — CHANGES_REQUESTED** [review]

File: frontend/src/entities/jobs/hpo/create/training-definitions/input-data-configuration.tsx [hunk 2]
```diff
@@ -151,6 +152,13 @@ export function Channel (props: ChannelProps) {
      *   .DataSource.S3DataSource.S3Uri
      */
 
+    let selectableItemsTypes: DatasetResourceSelectorSelectableItems[] = ['prefixes', 'objects'];
+    if (item.DataSource.S3DataSource?.S3DataType === S3DataType.S3Prefix) {
+        selectableItemsTypes = ['prefixes'];
+    } else if ([S3DataType.AugmentedManifestFile, S3DataType.ManifestFile].map(String).includes(item.DataSource.S3DataSource?.S3DataType ?? '')) {
```
[the remark this answers is not in the document]
Comment (dustins) [reply frontend/src/entities/jobs/hpo/create/training-definitions/input-data-configuration.tsx:158]: I think we do need this or at least something similar. Without it the linter assumes a `S3DataType[]` which only allows an `S3DataType` in the `include` method. There are some other options though. I'd say option 2 is cleaner?

```typescript
// options 1) declare the array above with the types defined
const  s3DataTypes: (string | undefined)[] = [S3DataType.AugmentedManifestFile, S3DataType.ManifestFile];

// option 2) lie to the linter
[S3DataType.AugmentedManifestFile, S3DataType.ManifestFile].includes(item.DataSource.S3DataSource?.S3DataType as S3DataType);
```

File: frontend/src/entities/jobs/transform/create/transform-create.tsx [hunk 3]
```diff
@@ -146,6 +147,13 @@ export function TransformCreate () {
         formSubmitting: false as boolean,
     });
 
+    let selectableItemsTypes: DatasetResourceSelectorSelectableItems[] = ['prefixes', 'objects'];
+    if (state.form.TransformInput.DataSource.S3DataSource.S3DataType === 'S3Prefix') {
+        selectableItemsTypes = ['prefixes'];
+    } else if (state.form.TransformInput.DataSource.S3DataSource.S3DataType === 'ManifestFile') {
```
[the remark this answers is not in the document]
Comment (dustins) [reply frontend/src/entities/jobs/transform/create/transform-create.tsx:153]: - We only give an option for AugmentedManifest files in Training/HPO.
  - Good call, yes we can/should replace these strings with an enum/union
- I don't believe we need other changes for the other job types
  - GT labeling jobs only allow a manifest file as input and you don't have to define a type
  - HPO/Training jobs both use the same `input-data-configuration.tsx` component

File: frontend/src/modules/dataset/dataset-selector.tsx [hunk 4]
```diff
@@ -25,29 +25,85 @@ import { useUsername } from '../../shared/util/auth-utils';
...
+                        scope += `/${username}`;
+                        break;
+                    case DatasetType.PROJECT:
+                        scope += `/${projectName}`;
+                        break;
+                }
+    
+                let isEmpty = true;
+                if (isFulfilled(response)) {
+                    const location = [scope, 'datasets', datasetContext.name, `${datasetContext.location}`].filter(Boolean).join('/');
```
[the remark this answers is not in the document]
Comment (dustins) [reply frontend/src/modules/dataset/dataset-selector.tsx:74]: `datasetContext` won't ever be undefined but `datasetContext.location` can be if the URI was for the top level of a dataset.

File: frontend/src/modules/dataset/dataset-selector.tsx (hunk 1 above)
Comment (dustins) [reply frontend/src/modules/dataset/dataset-selector.tsx:129]: Thanks. I wasn't really sure what to make this wording. If that is what we're doing with I can condense all this text into a single blurb.

File: frontend/src/modules/dataset/dataset-selector.tsx [hunk 5]
```diff
@@ -25,29 +25,85 @@ import { useUsername } from '../../shared/util/auth-utils';
...
+            })).then((response) => {
+                let scope = String(datasetContext.type);
+                switch (datasetContext.type) {
+                    case DatasetType.PRIVATE:
+                        scope += `/${username}`;
+                        break;
+                    case DatasetType.PROJECT:
+                        scope += `/${projectName}`;
+                        break;
+                }
```
[the remark this answers is not in the document]
Comment (dustins) [reply frontend/src/modules/dataset/dataset-selector.tsx:70]: I like that

File: frontend/src/entities/jobs/hpo/create/training-definitions/input-data-configuration.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (dustins) [reply frontend/src/entities/jobs/hpo/create/training-definitions/input-data-configuration.tsx:158]: None of those options were great, here is something I think is a bit cleaner.

```typescript
    let selectableItemsTypes: DatasetResourceSelectorSelectableItems[] = ['prefixes', 'objects'];
    switch (item.DataSource.S3DataSource?.S3DataType) {
        case S3DataType.S3Prefix:
            selectableItemsTypes = ['prefixes'];
            break;
        case S3DataType.ManifestFile:
        case S3DataType.AugmentedManifestFile:
            selectableItemsTypes = ['objects'];    
    }
```

**Review by douglas1850 — APPROVED** [review]

**Review by KyleLilly — APPROVED** [review]

**Pull request merged**